### ISSUE TITLE
Parsing of template args in single-quotes is incorrect.

### ISSUE THREAD

**Issue by skyrich62** [doxygen/doxygen#7302]
Consider the following code:
```c++
/// @file
/// Something

template <char C>
struct one { };

/// Less than
struct lt : one<'<'> { };

/// Equal
struct eq : one<'='> { };

/// Greater than
struct gt : one<'>'> { };

/// Left paren
struct lp : one<'('> { };

/// Decimal point.
struct pt : one<'.'> { };

/// Right paren
struct rp : one<')'> { };
```
The parser incorrectly considers '<', (in struct lt), to be the start of some construct, rather than a simple char.  Parsing resumes at struct gt, where it finds '>'.  Similarly, '(' and ')' cause problems as do '[' and ']'.  This causes missing documentation for everything between struct lt and struct gt, (i.e. there is no documentation for struct eq nor for struct gt). Addionally, the HAVE_DOT diagrams look horrible:

![structlt__inherit__graph](https://user-images.githubusercontent.com/23705081/66352493-74d14400-e914-11e9-8b19-72a791b71a6b.png)

This may be related to #6802. 

### Warning, Speculation follows
It seems to me that the parser/scanner is counting matching things, like <, >, (, ), [, ], etc, and failing to recognize context, (such as in a character literal).

### Workaround:  (But it's a bit ugly and kludgy)
```c++
#define LT '<'
#define GT '>'
#define LP '('
#define RP ')'

struct lt : one<LT> { }; 
   ... etc
```
*NOTE* using: `const auto LT = '<';` also works, and is slightly less kludgy.

### Reference
I ran into this while writing a grammar for [taocpp/PEGTL](https://github.com/taocpp/PEGTL).

**Comment by albert-github** [comment]
In your text you mention that you used `HAVE_DOT`, when not using dot but the doxygen 'internal' image generator we get a message like:
```
Error: /undefined in '
Operand stack:
   (one { };struct pt :one<'.'> { };struct rp :one<')
Execution stack:
   %interp_exit   .runexec2   --nostringval--   --nostringval--   --nostringval--   2   %stopped_push   --nostringval--   --nostringval--   --nostringval--   false   1   %stopped_push   2015   1   3
 %oparray_pop   2014   1   3   %oparray_pop   --nostringval--   1998   1   3   %oparray_pop   1884   1   3   %oparray_pop   --nostringval--   %errorexec_pop   .runexec2   --nostringval--   --nostringv
al--   --nostringval--   2   %stopped_push   --nostringval--
Dictionary stack:
   --dict:969/1684(ro)(G)--   --dict:0/20(G)--   --dict:114/200(L)--
Current allocation mode is local
Last OS error: No such file or directory
MiKTeX GPL Ghostscript 9.25: Unrecoverable error, exit code 1
```

To be certain about the settings and version you used:
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem?  Please don't add external links as they might not be persistent.
- Please also specify the doxygen version used.

**Comment by albert-github** [comment]
Problems also occur with double quotes, (escaped) single quotes.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #7303

**Comment by skyrich62** [comment]
$ doxygen --version
1.8.16
[7202.tar.gz](https://github.com/doxygen/doxygen/files/3703756/7202.tar.gz)

**Comment by albert-github** [comment]
Might be that the message is suppressed, tough you get an image (completely wrong of course).
I think the new version should be OK. I just tried with `HAVE_DOT=NO` and still get some warnings though the image is OK but this a different problem / issue and to be considered as a separate problem (probably to do with not escaping the round brackets).

**Comment by skyrich62** [comment]
I've confirmed that pull request 7303 solves the problem with HAVE_DOT=NO and HAVE_DOT=YES.

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.17.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by skyrich62** [comment]
Confirmed: Release 1.8.17 contains the fix for this issue.  
THANKS!   I've been using doxygen for about 20 years.  Love it.
